dds/DCPS/DataReaderImpl.h

```diff
@@ -434,6 +436,9 @@ class OpenDDS_Dcps_Export DataReaderImpl
   /// Release the instance with the handle.
   void release_instance(DDS::InstanceHandle_t handle);
 
+  // Take appropriate actions upon learning instance or view state has been updated
+  void state_updated(DDS::InstanceHandle_t handle);
```

Comment:
Given the current way DataReaderImpl and associates is implemented, this is the most straightforward way to implement this functionality.

However, it also points out why the design/implementation is bad.  A thread starting in DataReaderImpl or DataReaderImpl_T accesses SubscriptionInstance (which lacks encapsulation) which dispatches to InstanceState which dispatches back to DataReaderImpl.  Hence, we have indirect recursion.  This results in a number of undesirable things.  First, it bloats the public interface of the DataReader.  Alternatively, one must add friends which is similarly bad.  Second, it makes the code difficult to understand, e.g., if the recursion happens while iterating over a data structure, will the iterators become invalid?  Third, it makes the code difficult to understand with respect to concurrency. This leads to comments like "Why is this lock recursive?"